hal/src/nRF52840/exflash_hal.cpp

```diff
@@ -0,0 +1,717 @@
+    // Enter critical section
+    os_thread_scheduling(false, nullptr);
+
+    // Run QSPI TASK_ACTIVATE task as a workaround for the anomaly 215
+    exflash_qspi_activate();
+#endif // MODULE_FUNCTION != MOD_FUNC_BOOTLOADER
+
+    // This may block for up to 1ms (QSPI_DEF_WAIT_TIME_US * QSPI_DEF_WAIT_ATTEMPTS)
+    nrfx_err_t err = nrfx_qspi_cinstr_xfer(p_config, p_tx_buffer, p_rx_buffer);
+
+#if MODULE_FUNCTION != MOD_FUNC_BOOTLOADER
+    // Exit critical section
+    os_thread_scheduling(true, nullptr);
+#endif // MODULE_FUNCTION != MOD_FUNC_BOOTLOADER
+    return nrf_system_error(err);
+}
+
+int exflash_qspi_cinstr_quick_send(uint8_t opcode, nrf_qspi_cinstr_len_t length, void const* p_tx_buffer) {
+    nrf_qspi_cinstr_conf_t config = NRFX_QSPI_DEFAULT_CINSTR(opcode, length);
+    return exflash_qspi_cinstr_xfer(&config, p_tx_buffer, nullptr);
+}
+
+int perform_write(uintptr_t addr, const uint8_t* data, size_t size) {
+    return nrf_system_error(nrfx_qspi_write(data, size, addr));
+}
+
+nrf_qspi_cinstr_len_t get_nrf_cinstr_length(size_t length) {
+    nrf_qspi_cinstr_len_t cinstrLen = NRF_QSPI_CINSTR_LEN_1B;
+    switch (length) {
+        case 1: cinstrLen = NRF_QSPI_CINSTR_LEN_1B; break;
+        case 2: cinstrLen = NRF_QSPI_CINSTR_LEN_2B; break;
+        case 3: cinstrLen = NRF_QSPI_CINSTR_LEN_3B; break;
+        case 4: cinstrLen = NRF_QSPI_CINSTR_LEN_4B; break;
+        case 5: cinstrLen = NRF_QSPI_CINSTR_LEN_5B; break;
+        case 6: cinstrLen = NRF_QSPI_CINSTR_LEN_6B; break;
+        case 7: cinstrLen = NRF_QSPI_CINSTR_LEN_7B; break;
+        case 8: cinstrLen = NRF_QSPI_CINSTR_LEN_8B; break;
+        case 9: cinstrLen = NRF_QSPI_CINSTR_LEN_9B; break;
+        default: break;
+    }
+    return cinstrLen;
+}
+
+int mx25_enter_secure_otp() {
+    return exflash_qspi_cinstr_quick_send(HAL_QSPI_CMD_MX25_ENSO, get_nrf_cinstr_length(1), nullptr);
+}
+
+int mx25_exit_secure_otp() {
+    return exflash_qspi_cinstr_quick_send(HAL_QSPI_CMD_MX25_EXSO, get_nrf_cinstr_length(1), nullptr);
+}
+
+int mx25r64_switch_to_qspi() {
+    nrf_qspi_cinstr_conf_t cinstr_cfg = PRTCL_QSPI_DEFAULT_CINSTR(HAL_QSPI_CMD_STD_WRSR, NRF_QSPI_CINSTR_LEN_4B);
+    // Status register, configuration register 1, configuration register 2 values
+    uint8_t config_values[3] = { 0x40, 0x00, 0x02 };
+    return exflash_qspi_cinstr_xfer(&cinstr_cfg, config_values, nullptr);
+}
+
+int mx25l32_switch_to_qspi() {
+    nrf_qspi_cinstr_conf_t cinstr_cfg = PRTCL_QSPI_DEFAULT_CINSTR(HAL_QSPI_CMD_STD_WRSR, NRF_QSPI_CINSTR_LEN_2B);
+    uint8_t status_register[1] = { 0x40 };
+    return exflash_qspi_cinstr_xfer(&cinstr_cfg, status_register, nullptr);
+}
+
+int gd25_switch_to_qspi() {
+    nrf_qspi_cinstr_conf_t cinstr_cfg = PRTCL_QSPI_DEFAULT_CINSTR(HAL_QSPI_CMD_STD_WRSR, NRF_QSPI_CINSTR_LEN_2B);
+    uint8_t status_registers[3] = {0x00, 0x02, 0x00};
+
+    CHECK_FALSE(exflash_qspi_cinstr_xfer(&cinstr_cfg, &status_registers[0], nullptr), SYSTEM_ERROR_FLASH_IO);
```

Comment:
Why `CHECK_FALSE()` for a function returning `int` system error codes already? Same below.